### ISSUE TITLE
[Megatron] Task G: Wiring — connect core modules to run_pretrain + desloc_engine

### ISSUE THREAD

**Issue by dylanyunlon** [dylanyunlon/Neuron_SP#32]
## Problem

`run_pretrain.py` and `deepspeed/runtime/desloc_engine.py` expose a **wiring gap**: several core-module imports exist but are never exercised end-to-end, and the integration smoke test that would catch import-time or shape errors is missing.

Concrete symptoms observed in CI:
- `deepspeed.core.models` import failure → `run_pretrain.py` raises `RuntimeError` at module load time (correct behavior, but no CI test asserts this path is reachable and produces a meaningful error message).
- No test constructs `DesLocEngine` with a real `LlamaModel` (core backend) and verifies one forward+backward.

---

### Already wired (do NOT re-do)
| Module | File | Line(s) |
|--------|------|----------|
| `deepspeed.core.models.GPTModel` | `run_pretrain.py` | 223–227 |
| `deepspeed.core.transformer.TransformerConfig/Layer` | `run_pretrain.py` | 225 |
| `deepspeed.core.parallel_state` | `desloc_engine.py` | 46 |
| `deepspeed.core.distributed.{DistributedDataParallel, finalize_model_grads}` | `desloc_engine.py` | 56–59 |
| `deepspeed.core.optimizer.clip_grads.clip_grad_norm` | `desloc_engine.py` | 39 |
| `deepspeed.core.stream_manager.StreamManager` | `desloc_engine.py` | 51 |

### Already fixed in prod (do NOT re-do)
- **`run_pretrain.py` L236–257**: `NEURON_SP_ALLOW_FALLBACK_MODEL` guard + `RuntimeError` raise is fully implemented. Import failure causes a loud crash unless the env var is explicitly set to `"1"`. No code change needed here.
- **`desloc_engine.py` L566–569**: `hetero_bridge` import is intentionally a **hard import with no fallback** (see L566–568 comment and issue #13 no-fallback policy). `try/except ImportError` must NOT be added — silent fallback to plain AdamW caused weeks of wrong convergence.

### Remaining gap
1. **`LlamaModel.__init__` L398 / else-branch L427–437**: the fallback path (absolute pos-emb, no RoPE) is activated when `_HAS_CORE_MODELS=False`. No CI test asserts `_HAS_CORE_MODELS == True`, so a broken import would silently train the wrong architecture if `NEURON_SP_ALLOW_FALLBACK_MODEL=1` is set in CI.
2. **No integration test**: `tests/test_wiring_task_g.py` does not exist. `tests/test_run_pretrain.py` covers `LlamaModel` construction and forward on the fallback path, but does NOT assert `_HAS_CORE_MODELS`, does NOT construct `DesLocEngine`, and does NOT run a forward+backward through the engine.

---

### Fix 1 (sole code change) — new file `tests/test_wiring_task_g.py`

```python
import os, torch
os.environ.setdefault('WORLD_SIZE', '1')
os.environ.setdefault('RANK', '0')

from run_pretrain import LlamaModel, _HAS_CORE_MODELS
from deepspeed.runtime.desloc_engine import DesLocEngine, TrainingConfig


def test_core_model_wired():
    # _HAS_CORE_MODELS must be True; fallback env var must NOT be set in CI
    assert os.environ.get('NEURON_SP_ALLOW_FALLBACK_MODEL') != '1', (
        'NEURON_SP_ALLOW_FALLBACK_MODEL=1 is set in CI -- fallback arch used silently'
    )
    assert _HAS_CORE_MODELS, (
        'deepspeed.core.models.GPTModel not importable -- Task G wiring broken'
    )


def test_desloc_engine_constructs():
    # DesLocEngine must construct without error (requires hetero_bridge installed)
    cfg = TrainingConfig(
        vocab_size=512, hidden_size=64, num_layers=2, num_heads=4,
        seq_len=16, total_steps=2, micro_batch_size=1, global_batch_size=1,
    )
    model = LlamaModel(vocab_size=512, hidden_size=64, num_layers=2, num_heads=4, seq_len=16)
    engine = DesLocEngine(config=cfg, model=model)
    assert engine.optimizer is not None, 'DesLocEngine.optimizer not set after construction'


def test_forward_backward_single_gpu():
    # One forward+backward must not raise and must return loss > 0
    cfg = TrainingConfig(
        vocab_size=512, hidden_size=64, num_layers=2, num_heads=4,
        seq_len=16, total_steps=2, micro_batch_size=1, global_batch_size=1,
    )
    model = LlamaModel(vocab_size=512, hidden_size=64, num_layers=2, num_heads=4, seq_len=16)
    engine = DesLocEngine(config=cfg, model=model)
    ids = torch.randint(0, 512, (1, 16))
    lbl = torch.randint(0, 512, (1, 16))
    loss, scaled = engine.forward(ids, lbl, num_microbatches=1)
    scaled.backward()
    assert loss.item() > 0, f'loss={loss.item()} -- forward pass returned zero or negative loss'
```

---

## Files to Change

| File | Lines | Change |
|------|-------|--------|
| `tests/test_wiring_task_g.py` | new file | Integration smoke tests (3 functions above) |

No changes to `run_pretrain.py` or `desloc_engine.py` — both are already correctly implemented.

---

## Acceptance Criteria

- [ ] `python -c "from run_pretrain import _HAS_CORE_MODELS; assert _HAS_CORE_MODELS"` exits 0 on dev box
- [ ] `python -c "from deepspeed.runtime.desloc_engine import DesLocEngine"` exits 0
- [ ] `pytest tests/test_wiring_task_g.py::test_core_model_wired -v` passes — `_HAS_CORE_MODELS is True`
- [ ] `pytest tests/test_wiring_task_g.py::test_desloc_engine_constructs -v` passes — `engine.optimizer is not None`
- [ ] `pytest tests/test_wiring_task_g.py::test_forward_backward_single_gpu -v` passes — `loss.item() > 0`
- [ ] `pytest tests/test_run_pretrain.py tests/test_wiring_task_g.py -v` — no regression in existing suite
- [ ] `python run_pretrain.py --model-size 70m --steps 2` exits cleanly

---

## Dependencies

- Blocks: Task G completion sign-off
- Related: #13 (no-fallback policy — do NOT add try/except around `hetero_bridge` import in `desloc_engine.py`)
- Existing coverage: `tests/test_run_pretrain.py` covers LlamaModel fallback path; this issue adds the core-path + engine integration layer

**Comment by dylanyunlon** [comment]
## API Surface Inventory

### run_pretrain.py → core modules
| Call site | Current state | Status |
|-----------|--------------|--------|
| `from deepspeed.core.models import GPTModel` | Import exists (lines 224–227), try/except with `NEURON_SP_ALLOW_FALLBACK_MODEL` gate | ✅ wired |
| `from deepspeed.core.transformer import TransformerConfig, TransformerLayer` | Imported alongside GPTModel | ✅ wired |
| `LlamaModel.__init__` using `GPTModel` backend | `_HAS_CORE_MODELS` path builds `TransformerConfig`, calls `GPTModel(config, pre_process, post_process)` | ✅ wired |
| `GPTModel.forward` return shape | GPTModel returns `[B,T,vocab]` when `labels=None`; `run_desloc` passes model to DesLocEngine, not calling forward directly | ✅ correct |
| `parallel_state.initialize_model_parallel()` in `run_desloc` | **MISSING** — `dist.init_process_group` is called but `initialize_model_parallel(tensor_model_parallel_size=1, pipeline_model_parallel_size=1)` is never called before `DesLocEngine.__init__` | ❌ gap |
| `core.tensor_parallel.random.model_parallel_cuda_manual_seed` | Called in `run_desloc` as RNG seed before model creation | ✅ wired |

### desloc_engine.py → core modules
| Call site | Current state | Status |
|-----------|--------------|--------|
| `import deepspeed.core.parallel_state as parallel_state` | Top-level import | ✅ wired |
| `from deepspeed.core.distributed import DistributedDataParallel, finalize_model_grads` | Top-level import | ✅ wired |
| `from deepspeed.core.optimizer.clip_grads import clip_grad_norm` | Top-level import | ✅ wired |
| `from deepspeed.core.hetero_bridge import engine_integration` | Called in `__init__` Phase 5 | ✅ wired |
| `DesLocEngine.forward()` → `logits = self.model(input_ids)` | Returns `[B, T, vocab]` — loss computed as `F.cross_entropy(logits[:,:-1,:], labels[:,:T-1])` | ✅ correct for GPTModel (labels=None path) |
| `DesLocEngine.forward()` layer_log path | Hard-coded MiniTransformer attribute check (`embedding/blocks/norm/lm_head`) — does NOT work for GPTModel (`decoder/embedding.word_embeddings`) | ⚠️ minor — only when `NEURON_SP_LAYER_LOG=1` |
| `_get_model_layers()` | Checks `m._gpt.decoder.layers` (GPTModel path) at line 1442 | ✅ wired |

---

## Gap Analysis

### Gap 1 (Critical): `parallel_state.initialize_model_parallel()` not called in `run_desloc`

**Root cause:** `run_desloc()` calls `dist.init_process_group()` then immediately builds `LlamaModel` + `DesLocEngine`. `DesLocEngine.__init__` calls `hetero_bridge.install()` → `TierMap.discover()` → reads `parallel_state.is_initialized()`. Because `initialize_model_parallel()` was never called, `parallel_state._DATA_PARALLEL_GROUP` etc. are `None`, so every `parallel_state.get_data_parallel_*()` call falls through to the `dist.get_*()` fallback. This is functionally OK for TP=PP=1 but causes the `initialize_model_parallel()` guard in `parallel_state.is_initialized()` to return `False`, disabling the DP group path in `finalize_model_grads` and `CoreDDP`.

**Fix:** After `dist.init_process_group(...)`, before model creation, call:
```python
parallel_state.initialize_model_parallel(
    tensor_model_parallel_size=getattr(tc, "tensor_parallel_size", 1),
    pipeline_model_parallel_size=getattr(tc, "pipeline_parallel_size", 1),
)
```

### Gap 2 (Minor): `NEURON_SP_LAYER_LOG=1` layer-by-layer forward does not handle GPTModel

The structured forward path in `DesLocEngine.forward()` checks for `self.model.embedding` + `self.model.blocks` (MiniTransformer layout). GPTModel uses `self.model._gpt.embedding` + `self.model._gpt.decoder`. The fallback opaque call works, but the structured debug path is blind when GPTModel is active.

**Fix:** Add GPTModel-aware branch in `_has_structured_fwd` check.

### Gap 3 (Verified OK): `LlamaModel` forward with GPTModel backend

When `_HAS_CORE_MODELS=True`, `LlamaModel.forward()` calls `self._gpt(input_ids, position_ids, attention_mask=None, labels=None)` and gets `[B, T, vocab]`. `DesLocEngine.forward()` then reshapes as `logits[:, :-1, :]` — matching the shift-1 label convention. No shape mismatch.

---

## Before / After

### Before (current state)
```
run_desloc():
  dist.init_process_group(nccl)          # ✅
  # parallel_state.initialize_model_parallel()  ← MISSING
  LlamaModel(...)                         # builds GPTModel via _HAS_CORE_MODELS
  DesLocEngine.__init__():
    parallel_state.is_initialized() → False   # always falls to dist.get_*() fallback
    hetero_bridge.install()                    # TierMap uses fallback rank path
    CoreDDP init: skips parallel_state DP group
    finalize_model_grads: skips parallel_state DP group → uses plain model
```

### After (this PR)
```
run_desloc():
  dist.init_process_group(nccl)          # ✅
  parallel_state.initialize_model_parallel(tp=1, pp=pp_size)  # ← NEW
  LlamaModel(...)                         # unchanged
  DesLocEngine.__init__():
    parallel_state.is_initialized() → True    # ✅ enables DP group paths
    CoreDDP gets real DP group from parallel_state
    finalize_model_grads gets real DP group
    layer_log: GPTModel-aware structured forward path
```

## Changes
- `run_pretrain.py` `run_desloc()`: insert `parallel_state.initialize_model_parallel()` after `dist.init_process_group`, using `tensor_parallel_size` and `pipeline_parallel_size` from `tc` (TrainingConfig)
- `deepspeed/runtime/desloc_engine.py` `forward()`: extend `_has_structured_fwd` to detect GPTModel backend
